src/bin/pg_autoctl/watch.c

```diff
@@ -0,0 +1,1102 @@
+/*
+ * src/bin/pg_autoctl/watch.c
+ *     Implementation of a CLI to show events, states, and URI from the
+ *     pg_auto_failover monitor.
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the PostgreSQL License.
+ *
+ */
+#include <inttypes.h>
+#include <getopt.h>
+#include <unistd.h>
+
+#include <curses.h>
+#include <ncurses.h>
+
+#include <sys/ioctl.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+#include "postgres_fe.h"
+
+#include "cli_common.h"
+#include "commandline.h"
+#include "defaults.h"
+#include "env_utils.h"
+#include "ipaddr.h"
+#include "keeper_config.h"
+#include "keeper.h"
+#include "monitor_config.h"
+#include "monitor_pg_init.h"
+#include "monitor.h"
+#include "nodestate_utils.h"
+#include "parsing.h"
+#include "pgctl.h"
+#include "pghba.h"
+#include "pgsetup.h"
+#include "pgsql.h"
+#include "pidfile.h"
+#include "state.h"
+#include "string_utils.h"
+#include "watch.h"
+#include "watch_colspecs.h"
+
+volatile sig_atomic_t window_size_changed = 0;      /* SIGWINCH */
+
+static int print_watch_header(WatchContext *context, int r);
+static int print_watch_footer(WatchContext *context);
+static int print_nodes_array(WatchContext *context, int r, int c);
+static int print_events_array(WatchContext *context, int r, int c);
+
+static ColPolicy * pick_column_policy(WatchContext *context);
+static bool compute_column_spec_lens(WatchContext *context);
+static int compute_column_size(ColumnType type, NodeAddressHeaders *headers);
+
+
+static void print_column_headers(WatchContext *context,
+								 ColPolicy *policy,
+								 int r,
+								 int c);
+
+static void print_node_state(WatchContext *context,
+							 ColPolicy *policy,
+							 int index,
+							 int r,
+							 int c);
+
+static void print_events_headers(WatchContext *context, int r, int c);
+
+static void watch_set_state_attributes(NodeState state, bool toggle);
+
+static void clear_line_at(int row);
+
+/*
+ * catch_sigwinch is registered as the SIGWINCH signal handler.
+ */
+static void
+catch_sigwinch(int sig)
+{
+	window_size_changed = 1;
+	pqsignal(sig, catch_sigwinch);
+}
```

Comment:
Probably nice to move this to signals.c, even though it's only used here.

Reply:
I don't feel like this would help clarify the code, so I didn't do that yet. Let's chat about it if you feel more strongly about it?